File: src/services/geminiService.ts
import { format } from 'date-fns';
import { ru } from 'date-fns/locale';
import { GoogleGenerativeAI } from "@google/generative-ai";

const isVercel = window.location.hostname !== 'localhost' && window.location.hostname !== '127.0.0.1';
const BASE_URL = isVercel ? '' : `http://${window.location.hostname}:8001`;

// Инициализация Gemini API для прямого доступа (используется если бэкенд недоступен)
const API_KEY = import.meta.env.VITE_GEMINI_API_KEY;
if (!API_KEY && isVercel) {
  console.error("КРИТИЧЕСКАЯ ОШИБКА: VITE_GEMINI_API_KEY не найден в переменных окружения Vercel!");
}
const genAI = API_KEY ? new GoogleGenerativeAI(API_KEY) : null;

// Хелпер для конвертации файла в формат Gemini
async function fileToGenerativePart(file: File): Promise<{ inlineData: { data: string; mimeType: string } }> {
  const base64EncodedDataPromise = new Promise<string>((resolve) => {
    const reader = new FileReader();
    reader.onloadend = () => resolve((reader.result as string).split(',')[1]);
    reader.readAsDataURL(file);
  });
  return {
    inlineData: {
      data: await base64EncodedDataPromise,
      mimeType: file.type,
    },
  };
}

/**
 * Сервис для анализа видео через локальный бэкенд
 */

export const analyzeVideo = async (videoSource: string | File, isUrl: boolean = true): Promise<{ summary: string; title: string }> => {
  if (!isUrl) {
    // Демо-режим для локальных файлов (загрузка через браузер)
    await new Promise(resolve => setTimeout(resolve, 2000));
    const title = (videoSource as File).name;
    const summary = `# 📁 Конспект файла: ${title}\n\n## 🎯 Анализ загруженного видео завершен.\n\n*(В этой версии анализ локальных файлов работает в режиме демонстрации)*`;
    return { summary, title };
  }

  // Если мы на Vercel, обработка видео (транскрибация) через этот метод недоступна без бэкенда
  if (isVercel) {
    throw new Error("Анализ видео временно недоступен на Vercel (требуется Python-сервер для транскрибации).");
  }

  try {
    // Используем наш бэкенд для транскрибации и суммаризации
    const response = await fetch(`${BASE_URL}/summarize`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ url: videoSource }),
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.detail || 'Ошибка при обработке видео на бэкенде');
    }

    const data = await response.json();
    
    // Возвращаем структурированный конспект и заголовок
    return {
      summary: data.summary,
      title: data.title || "Конспект видео"
    };
  } catch (error: any) {
    console.error("Error in analyzeVideo:", error);
    
    if (error.message?.includes("Failed to fetch")) {
      throw new Error(`Не удалось подключиться к серверу для обработки видео. На Vercel эта функция требует запущенного бэкенда или использования YouTube URL (если настроено).`);
    }
    
    throw new Error(error.message || "Произошла ошибка при обработке видео на сервере.");
  }
};

export const chatWithAI = async (message: string, history: any[] = []) => {
  // Пытаемся вызвать Gemini напрямую, если есть API ключ (для Vercel)
  if (genAI) {
    try {
      const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" });
      const chat = model.startChat({
        history: history.length > 0 ? history : [],
      });
      const result = await chat.sendMessage(message);
      return result.response.text();
    } catch (directError: any) {
      console.warn("Direct Gemini API call failed, falling back to backend:", directError);
      // Если ошибка в ключе или лимитах, пробрасываем её
      if (directError.message?.includes("API_KEY_INVALID") || directError.message?.includes("quota")) {
        throw directError;
      }
    }
  }

  // Если мы на Vercel и нет ключа, даже не пытаемся стучаться на localhost
  if (isVercel && !genAI) {
    throw new Error("ИИ недоступен: Добавьте VITE_GEMINI_API_KEY в настройки Vercel.");
  }

  try {
    const response = await fetch(`${BASE_URL}/chat`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ message, history }),
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.detail || 'Ошибка при общении с ИИ');
    }

    const data = await response.json();
    return data.response;
  } catch (error: any) {
    console.error("Error in chatWithAI:", error);
    if (error.message?.includes("Failed to fetch")) {
      throw new Error("Не удалось подключиться к ИИ. Убедитесь, что бэкенд запущен или добавлен VITE_GEMINI_API_KEY.");
    }
    throw new Error(error.message || "Не удалось получить ответ от ИИ.");
  }
};

/**
 * Распознает расписание по изображению через бэкенд
 */
export const recognizeScheduleFromImage = async (imageFile: File, group: string = ""): Promise<PlanItem[]> => {
  // Пытаемся использовать прямой API Gemini если есть ключ
  if (genAI) {
    try {
      const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" });
      const imageData = await fileToGenerativePart(imageFile);
      
      const group_focus = group.includes("2") 
        ? "ВНИМАНИЕ: Пользователь из 2 ГРУППЫ. Игнорируй левую колонку, бери данные ТОЛЬКО из ПРАВОЙ колонки."
        : "ВНИМАНИЕ: Пользователь из 1 ГРУППЫ. Бери данные ТОЛЬКО из ЛЕВОЙ колонки.";

      const prompt = `
        Проанализируй это изображение расписания. Группа: ${group}.
        ${group_focus}
        Извлеки предметы (title), время (start, end), кабинет (room) и день недели (day).
        Верни ТОЛЬКО массив JSON в формате:
        [{"title": "...", "start": "HH:mm", "end": "HH:mm", "room": "...", "day": "понедельник"}]
      `;

      const result = await model.generateContent([prompt, imageData]);
      const response = result.response.text();
      const cleanJson = response.replace(/```json/g, '').replace(/```/g, '').trim();
      const schedule = JSON.parse(cleanJson);
      
      return schedule.map((item: any) => ({
        ...item,
        type: 'school',
        isRecommendation: false
      }));
    } catch (directError) {
      console.warn("Direct Gemini vision API call failed, falling back to backend:", directError);
    }
  }

  // Если мы на Vercel и нет ключа, даже не пытаемся стучаться на localhost
  if (isVercel && !genAI) {
    throw new Error("Распознавание недоступно: Добавьте VITE_GEMINI_API_KEY в настройки Vercel.");
  }

  const formData = new FormData();
  formData.append('file', imageFile);
  formData.append('group', group);

  try {
    const response = await fetch(`${BASE_URL}/recognize-schedule`, {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.detail || 'Ошибка при распознавании расписания');
    }

    const data = await response.json();
    return data.schedule.map((item: any) => ({
      start: item.start,
      end: item.end,
      title: item.title,
      room: item.room,
      day: item.day,
      type: 'school',
      isRecommendation: false
    }));
  } catch (error: any) {
    console.error("Error in recognizeScheduleFromImage:", error);
    if (error.message?.includes("Failed to fetch")) {
      throw new Error("Не удалось подключиться к серверу распознавания. Добавьте VITE_GEMINI_API_KEY для прямой работы без бэкенда.");
    }
    throw new Error(error.message || "Не удалось распознать расписание.");
  }
};

export interface PlanItem {
  start: string;
  end: string;
  title: string;
  type: 'rest' | 'productivity' | 'activity' | 'routine' | 'school' | 'sleep' | 'meal';
  isRecommendation: boolean;
  room?: string;
  day?: string;
}

export interface ScheduleAnalysis {
  analysis: string;
  plan: PlanItem[];
}

export const analyzeSchedule = async (date: string, schedule: any[], settings: any): Promise<ScheduleAnalysis> => {
  const dateObj = new Date(date);
  const formattedDateForAI = format(dateObj, 'd MMMM', { locale: ru });
  const dayName = format(dateObj, 'EEEE', { locale: ru });
  const isWeekend = dateObj.getDay() === 0 || dateObj.getDay() === 6;

  const prompt = `
    Проанализируй мой график на ${formattedDateForAI} (${dayName}) и составь ПОЛНЫЙ идеальный план дня.
    
    ВАЖНО: Следующие активности являются ШАБЛОНОМ. Они ДОЛЖНЫ быть в плане каждый день строго в указанное время:
    - Подъем: ${settings.wakeUpTime}
    - Отбой: ${settings.bedTime}
    - Завтрак: ${settings.breakfastTime}
    - Обед: ${settings.lunchTime}
    - Ужин: ${settings.dinnerTime}
    
    ШКОЛА:
    ${isWeekend 
      ? "- Сегодня ВЫХОДНОЙ (суббота или воскресенье), поэтому ШКОЛЫ И ПУТИ ДО НЕЕ НЕТ. НЕ ВКЛЮЧАЙ школу в план." 
      : `- Школа: ${settings.schoolStart} - ${settings.schoolEnd} (обязательно включи время в пути по ${settings.commuteTime} мин до и после)`}

    Мои постоянные активности (шаблон):
    ${(() => {
      const dayNum = dateObj.getDay(); // 0-6, 0 is Sunday
      const ruDayNum = dayNum === 0 ? 7 : dayNum; // Convert to 1-7 (Mon-Sun)
      
      const relevantRoutines = (settings?.routineActivities || []).filter((a: any) => 
        !a.days || a.days.length === 0 || a.days.includes(ruDayNum)
      );
      
      return relevantRoutines.length > 0 
        ? relevantRoutines.map((a: any) => `- ${a.startTime} - ${a.endTime}: ${a.title}`).join('\n')
        : 'Нет дополнительных постоянных активностей на этот день';
    })()}

    Мои специфические активности на этот конкретный день (из расписания):
    ${(schedule || []).length > 0 ? schedule.map(e => `- ${e.startTime} - ${e.endTime}: ${e.title} (${e.type})`).join('\n') : 'Нет специфических активностей'}

    Твоя задача:
    1. Составь последовательный план дня с момента подъема до отбоя.
    2. Используй ШАБЛОН выше (подъем, еда, отбой) как незыблемую основу. Они ДОЛЖНЫ быть в плане каждый день, включая выходные.
    3. ОБЯЗАТЕЛЬНО ВЫДЕЛИ МИНИМУМ 2 ЧАСА В ДЕНЬ НА ДОМАШНЮЮ РАБОТУ (Homework). 
       - Разбей это время на 2 отдельных блока по 1 часу каждый.
       - Распредели эти блоки гармонично в течение дня (например, один днем после школы/обеда, другой вечером перед ужином), чтобы не перегружать меня.
       - Используй type: "productivity" и title: "Домашняя работа" для этих блоков.
    4. ${isWeekend 
        ? "В плане НЕ ДОЛЖНО быть школы и времени в пути. Заполни это время отдыхом или полезными делами." 
        : 'ОБЯЗАТЕЛЬНО объедини все школьные уроки в один блок "Школа" с общим временем начала и конца. Включи время в пути.'}
    4. ВКЛЮЧИ в план все мои постоянные и специфические активности.
    5. Все шаблонные и мои активности должны иметь isRecommendation: false.
    6. ЗАПОЛНИ все свободное время (которого в выходные будет гораздо больше) интересными и полезными рекомендациями (isRecommendation: true) для отдыха, саморазвития или хобби. План должен быть насыщенным и полным, а не пустым!
    7. Дай общий краткий совет по продуктивности на этот день.

    ОТВЕТЬ СТРОГО В ФОРМАТЕ JSON. Не пиши ничего, кроме JSON. 
    ВАЖНО: Избегай специальных символов в строках (кавычки, обратные слеши). Если они необходимы, экранируй их правильно (\\).
    Пример структуры:
    {
      "analysis": "краткий текст анализа дня",
      "plan": [
        {
          "start": "08:00",
          "end": "08:30",
          "title": "Завтрак",
          "type": "meal",
          "isRecommendation": false
        }
      ]
    }
    
    Допустимые значения для "type": "rest", "productivity", "activity", "routine", "school", "sleep", "meal".
  `;

  try {
    const response = await chatWithAI(prompt);
    
    // Очистка ответа от возможных markdown-тегов ```json ... ```
    const cleanResponse = response
      .replace(/```json/g, '')
      .replace(/```/g, '')
      .trim();

    const jsonMatch = cleanResponse.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      console.error("AI Response was:", response);
      throw new Error("ИИ вернул некорректный формат данных");
    }
    
    let jsonStr = jsonMatch[0];
    
    // Очистка JSON от типичных ошибок ИИ
    jsonStr = jsonStr
      .replace(/,\s*([\]}])/g, '$1') // Удаление лишних запятых
      .replace(/\\(?!["\\\/bfnrtu])/g, '\\\\'); // Экранирование одиночных обратных слешей

    try {
      return JSON.parse(jsonStr);
    } catch (parseError) {
      // Попытка еще более агрессивной очистки если первый раз не вышло
      try {
        const ultraClean = jsonStr
          .replace(/[\u0000-\u001F\u007F-\u009F]/g, "") // Удаление невидимых управляющих символов
          .replace(/\\/g, "\\\\") // Экранируем ВСЕ слеши (рискованно, но может помочь)
          .replace(/\\\\"/g, "\\\""); // Возвращаем экранированные кавычки обратно
        return JSON.parse(ultraClean);
      } catch (e) {
        console.error("Failed to parse JSON even after cleaning:", jsonStr);
        console.error("Original response:", response);
        throw new Error("Ошибка при обработке плана дня. Попробуйте еще раз.");
      }
    }
  } catch (error) {
    console.error("Error in analyzeSchedule:", error);
    throw error;
  }
};

File: backend/services/gemini_service.py
import google.generativeai as genai
from backend.config import GEMINI_API_KEY

class GeminiService:
    def __init__(self):
        self.api_key = GEMINI_API_KEY
        self.models_priority = [
            'gemini-2.5-flash-lite',
            'gemini-2.5-flash',
            'gemini-2.0-flash-lite-preview-02-05',
            'gemini-2.0-flash'
        ]
        if self.api_key:
            genai.configure(api_key=self.api_key)
        else:
            print("ВНИМАНИЕ: GEMINI_API_KEY не установлен в .env")

    async def recognize_schedule_from_image(self, image_data: bytes, mime_type: str, group: str = ""):
        if not self.api_key:
            return None
        
        # Определяем, какую колонку искать на основе группы
        group_focus = ""
        is_group_2 = "2" in group
        
        if is_group_2:
            group_focus = """
            ВНИМАНИЕ: Пользователь из 2 ГРУППЫ. 
            Твоя задача — игнорировать левую колонку с предметами.
            Бери данные ТОЛЬКО из ПРАВОЙ колонки в каждой строке.
            Если в ячейке написано два предмета (например, через дробь / или в разных частях ячейки), бери тот, что СПРАВА.
            """
        else:
            group_focus = """
            ВНИМАНИЕ: Пользователь из 1 ГРУППЫ.
            Бери данные ТОЛЬКО из ЛЕВОЙ колонки в каждой строке.
            Если в ячейке написано два предмета, бери тот, что СЛЕВА.
            """

        prompt = f"""
        Проанализируй это изображение расписания занятий. 
        Группа пользователя: {group if group else "не указана"}.
        {group_focus}
        
        Извлеки список предметов на всю неделю (пн-пт).
        
        КРИТИЧЕСКИЕ ПРАВИЛА РАЗДЕЛЕНИЯ ГРУПП:
        1. Таблица разделена вертикально на две части для каждой строки.
        2. ЛЕВАЯ часть ячейки/строки = 1 ГРУППА.
        3. ПРАВАЯ часть ячейки/строки = 2 ГРУППА.
        4. Если ячейка цельная (на всю ширину) — этот предмет общий для обеих групп.
        5. ОШИБКА ЗАПРЕЩЕНА: Не перепутай колонки. Если пользователь во 2-й группе, ты ДОЛЖЕН проигнорировать первый (левый) предмет в строке и взять второй (правый).

        ПРАВИЛА ВРЕМЕНИ (СЕТКА ЗВОНКОВ):
        - Длительность одного урока: 40 минут.
        - Перемены между уроками:
          1. После 1-го урока: 5 минут
          2. После 2-го урока: 10 минут
          3. После 3-го урока: 10 минут
          4. После 4-го урока: 15 минут
          5. После 5-го урока: 10 минут
          6. После 6-го урока: 10 минут
        
        Если в расписании указан только номер урока (например, "1 урок") или только время начала, используй эту сетку для расчета точного времени начала (start) и конца (end).
        Обычно занятия начинаются в 08:00 или 08:30. Если время начала первого урока на фото другое, адаптируй всю сетку.
        
        СТРУКТУРА РАСПИСАНИЯ:
        1. Дни недели расположены горизонтальными блоками.
        
        2. Колонки групп (КРИТИЧЕСКИ ВАЖНО):
           - Каждая строка с предметами разделена на две основные колонки.
           - ЛЕВАЯ КОЛОНКА всегда соответствует "1 Группе" (10S-1, 9O-1 и т.д.).
           - ПРАВАЯ КОЛОНКА всегда соответствует "2 Группе" (10S-2, 9O-2 и т.д.).
           - Если пользователь выбрал "1 группа", бери данные ТОЛЬКО из ЛЕВОЙ колонки.
           - Если пользователь выбрал "2 группа", бери данные ТОЛЬКО из ПРАВОЙ колонки.
           - Если ячейка предмета одна на всю ширину (общая), бери её.
        
        3. Формат ячеек:
           - Название предмета и кабинет.
        
        ЗАДАЧА:
        1. Определи день недели.
        2. Выбери нужную колонку в зависимости от группы ({group}).
        3. Извлеки название предмета (title), время (start/end) и кабинет (room).
        
        Верни список предметов в формате JSON:
        [
          {{"title": "...", "start": "HH:mm", "end": "HH:mm", "room": "...", "day": "понедельник", "type": "school"}},
          ...
        ]
        Верни ТОЛЬКО массив JSON.
        """
        
        for model_name in self.models_priority:
            try:
                model = genai.GenerativeModel(model_name)
                response = await model.generate_content_async([
                    prompt,
                    {'mime_type': mime_type, 'data': image_data}
                ])
                
                # Очистка от markdown блоков если есть
                text = response.text.strip()
                if text.startswith("```json"):
                    text = text[7:-3].strip()
                elif text.startswith("```"):
                    text = text[3:-3].strip()
                
                import json
                return json.loads(text)
            except Exception as e:
                print(f"Ошибка Vision в модели {model_name}: {str(e)}")
                if "429" in str(e):
                    print(f"Лимит запросов для {model_name}")
                elif "403" in str(e):
                    print(f"Ошибка доступа/API ключа для {model_name}")
                continue
        
        return None

    async def get_response(self, message: str, history: list = []):
        if not self.api_key:
            return "GEMINI_API_KEY не настроен."

        for model_name in self.models_priority:
            try:
                model = genai.GenerativeModel(model_name)
                # Преобразуем историю в формат Google Generative AI
                chat_history = []
                for msg in history:
                    role = "user" if msg["role"] == "user" else "model"
                    
                    # Извлекаем текст из различных возможных форматов сообщения
                    content = ""
                    if isinstance(msg.get("content"), str):
                        content = msg["content"]
                    elif isinstance(msg.get("parts"), list):
                        # Собираем текст из всех частей
                        parts_texts = []
                        for p in msg["parts"]:
                            if isinstance(p, dict) and "text" in p:
                                parts_texts.append(p["text"])
                            elif isinstance(p, str):
                                parts_texts.append(p)
                        content = " ".join(parts_texts)
                    
                    if content:
                        chat_history.append({"role": role, "parts": [content]})
                
                chat = model.start_chat(history=chat_history)
                response = await chat.send_message_async(message)
                return response.text
            except Exception as e:
                print(f"Ошибка Chat в модели {model_name}: {str(e)}")
                if "429" in str(e):
                    print(f"Лимит запросов для {model_name}")
                elif "403" in str(e):
                    print(f"Ошибка доступа/API ключа для {model_name}")
                continue
        
        return "Извините, возникла ошибка при обработке запроса. Пожалуйста, попробуйте позже."

gemini_service = GeminiService()
